tests/test_distillation_loss.py

```diff
@@ -0,0 +1,185 @@
+def test_reverse_kl_no_tp(use_mask):
+    torch.manual_seed(0)
+    batch_size, seq_len, vocab_size = 2, 3, 5
+    logits = torch.randn(batch_size, seq_len, vocab_size, requires_grad=True)
+    target = torch.randn(batch_size, seq_len, vocab_size)
+    loss_mask = torch.tensor([[1.0, 1.0, 0.0], [1.0, 1.0, 0.0]]) if use_mask else None
+
+    loss, grad = reverse_kl_forward_backward(
+        logits=logits,
+        target=target,
+        loss_mask=loss_mask,
+        grad_output=1.0,
+        group=None,
+        target_format=TargetFormat.logits,
+        sequence_parallel_logits=False,
+    )
+    _assert_loss_and_grad(logits, loss, grad)
+
+    # Manual reference: sum over vocab then average over valid tokens.
+    teacher_log_probs = torch.log_softmax(target, dim=-1)
+    student_log_probs = torch.log_softmax(logits, dim=-1)
+    per_sample = torch.nn.functional.kl_div(
+        teacher_log_probs, student_log_probs, reduction="none", log_target=True
+    ).sum(dim=-1)
+    if loss_mask is not None:
+        per_sample = per_sample * loss_mask
+        valid_tokens = loss_mask.sum()
+    else:
+        valid_tokens = logits.shape[0] * logits.shape[1]
+    reference = per_sample.sum() / valid_tokens
+    torch.testing.assert_close(loss, reference, atol=1e-6, rtol=1e-6)
+
+
+def _vocab_tp_worker(rank: int, group: dist.ProcessGroup, use_mask: bool):
+    torch.manual_seed(0)
+    world_size = dist.get_world_size(group)
+
+    batch_size, seq_len, vocab_per_rank = 2, 3, 5
+    full_vocab = vocab_per_rank * world_size
+    full_logits = torch.randn(batch_size, seq_len, full_vocab)
+    full_target = torch.randn(batch_size, seq_len, full_vocab)
+    full_mask = torch.tensor([[1.0, 1.0, 0.0], [1.0, 1.0, 0.0]]) if use_mask else None
+
+    start = rank * vocab_per_rank
+    end = start + vocab_per_rank
+    logits = full_logits[:, :, start:end].clone().requires_grad_(True)
+    target = full_target[:, :, start:end].clone()
+    loss_mask = full_mask.clone() if full_mask is not None else None
+
+    loss, grad = reverse_kl_forward_backward(
+        logits=logits,
+        target=target,
+        loss_mask=loss_mask,
+        grad_output=None,
+        group=group,
+        target_format=TargetFormat.logits,
+        sequence_parallel_logits=False,
+    )
+    _assert_loss_and_grad(logits, loss, grad)
+
+    if rank == 0:
+        ref_loss, _ = reverse_kl_forward_backward(
+            logits=full_logits.clone(),
+            target=full_target.clone(),
+            loss_mask=full_mask.clone() if full_mask is not None else None,
+            grad_output=None,
+            group=None,
+            target_format=TargetFormat.logits,
+            sequence_parallel_logits=False,
+        )
+    else:
+        ref_loss = torch.zeros_like(loss)
+    dist.broadcast(ref_loss, src=0, group=group)
+    torch.testing.assert_close(loss, ref_loss, atol=1e-6, rtol=1e-6)
+
+
+def _ce_vocab_tp_worker(rank: int, group: dist.ProcessGroup, use_mask: bool):
```

Comment:
We might want to match the implementation and parametrization from `test_cross_entropy`